### ISSUE TITLE
Angular 12 and Storybook with Webpack 5 issues

### ISSUE THREAD

**Issue by vrapalis** [nrwl/nx#5688]
## Current Behavior

Can't run storybook based on nx angular, nx run shared-ui:storybook

## Expected Behavior

Run storybook on web-server

## Steps to Reproduce

npx create-nx-workspace@latest storybook-example
nx g library shared/ui --tag=scope:shared,type:ui
 nx g c nav --flat -s -t --skipTests --project=shared-ui
npm i @nrwl/storybook -D
nx g @nrwl/angular:storybook-configuration shared-ui
nx run shared-ui:storybook

This issue may not be prioritized if details are not provided to help us reproduce the issue.

### Failure Logs

![image](https://user-images.githubusercontent.com/38319324/118520142-15672700-b73a-11eb-9d0e-310e21d99b32.png)

### Environment

Node: v14.17.0
Npm: 7.13.0
![image](https://user-images.githubusercontent.com/38319324/118520335-46475c00-b73a-11eb-964f-99d6bf1d33a7.png)

**Comment by vrapalis** [comment]
https://github.com/storybookjs/storybook/issues/14909

**Comment by literalpie** [comment]
The first screenshot is showing babel warnings that shouldn't affect building or anything. You should be able to avoid them with `npm I -D @babel/preset-env@7.13.0`.

**Comment by kingjordan** [comment]
I am getting the exact same errors,

**REPRODUCE:**
I have a totally clean install of nx workspace

	1. npx create-nx-workspace test-org --preset=angular
	2. cd test-org
	3. npx ng generate lib ui
	4. npx ng g component header --project=ui --export
	5. npm i -D @nrwl/storybook
	6. nx g storybook-configuration ui --configureCypress=true --generateCypressSpecs=true --generateStories=true
	7. nx run ui:storybook

**RESULT:**
I get an error that says:

>  NX ERROR Running Target "ui:storybook" failed
> Failed Tasks
> - ui:storybook

![storybook](https://user-images.githubusercontent.com/5931768/118551795-569c0e80-b71b-11eb-8f82-b6b2679842a8.jpg)

I can use this header component in my test-app and it works.

I have tried creating fresh workspaces multiple ways and am getting this same error every time.

**What I have tried:**

1. I have tried creating a blank workspace and adding an angular project after and then adding story book -> same error
2. I have tried creating the workspace using `npx create-nx-workspace@latest` (test-org, angular, test-app, scss) -> same error
3. I have followed the directions exactly from [this tutorial](https://www.youtube.com/watch?v=cXOkmOy-8dk&t=47s) -> same error
4. I have tried updating @nrwl/storybook, @storybook/addon-knobs, and @storybook/builder-webpack5 from 6.2.7 to 6.2.9 -> same error


**NX REPORT**

  Node : 14.15.1
  OS   : win32 x64
  npm  : 7.13.0

  nx : Not Found
  @nrwl/angular : 12.3.3
  @nrwl/cli : 12.3.3
  @nrwl/cypress : 12.3.3
  @nrwl/devkit : 12.3.3
  @nrwl/eslint-plugin-nx : 12.3.3
  @nrwl/express : Not Found
  @nrwl/jest : 12.3.3
  @nrwl/linter : 12.3.3
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 12.3.3
  @nrwl/web : Not Found
  @nrwl/workspace : 12.3.3
  @nrwl/storybook : 12.3.3
  @nrwl/gatsby : Not Found
  typescript : 4.2.4

**Comment by kingjordan** [comment]
@literalpie thanks for the tip to remove those babel warnings!
however, 
> npm i -D @babel/preset-env@7.13.0
did not fix the error, storybook still does not run in new nx workspace

**Comment by juristr** [comment]
@kingjordan thanks for reporting. We're looking into this. Though this seems to be a Storybook + Angular 12 related thing as even with a plain Angular CLI project it doesn't work.

There's an open Storybook issue (https://github.com/nrwl/nx/issues/5688#issuecomment-842454587) which tracks the current Angular 12 status.

**Comment by kingjordan** [comment]
@juristr thanks for looking into this, the joys of having third party library dependencies 😜
let me know if you need any help testing as I am more than willing.

**Comment by acostaaluiz** [comment]
I am facing the same issue. is there any way to solve it? with an old version? thanks guys.

**Comment by juristr** [comment]
Hey all, looks like the latest Storybook 6.2.3 alpha fixes most of the issues. I created a repo that shows a working version of Nx 12 + Ng 12 + Webpack 5 and Storybook. Specifically this commit shows what to do to fix it:

https://github.com/juristr/nx-angular12-storybook-webpack5/commit/229a596cb234ede0d6f0fda549f9b2b824887892

I'll keep this issue open as a reference and for following along as the Storybook folks fix this and provide a proper release. Hopefully this unblocks you all for now

**Comment by literalpie** [comment]
@juristr can you double-check that building storybook works? `nx run shared-ui:build-storybook` gives me an error in your example project.

**Comment by juristr** [comment]
@MickL yep the storybook build still fails. The problem is that `@storybook/core-server` is still having a dependency on Webpack 4, thus in the `JavaScriptModulesPlugin.js` when doing an instance check those won't match, as Storybook passes in a Webpack 4 instance, comparing it to a Webpack 5 instance

![image](https://user-images.githubusercontent.com/542458/118776827-af3cea00-b888-11eb-8482-09322a737c72.png)

I'm looking into it

**Comment by juristr** [comment]
As a quick update. I'm in contact with the Storybook team and they're looking into it. This is a weird NPM package hoisting issue 🤷‍♂️

**Comment by kingjordan** [comment]
thanks @juristr
I can confirm that a nx workspace that did not work (6.2.7 and 6.2.9) will now run storybook if I make the changes you linked to in package.json (6.3.0-alpha etc...)
But as @literalpie  mentioned, the build of story book still fails.  For me this unblocks development for us as we don't need the build right now.
Again, thank you so much for your effort solving this issue!

**Comment by juristr** [comment]
@kingjordan cool. Yeah I know the Storybook folks is working on a fix so hopefully we have one in the next couple of days 🤞

**Comment by dean-g** [comment]
running to the same error.
![image](https://user-images.githubusercontent.com/12157452/119165742-6ec5a380-ba23-11eb-990c-b3f7d3235840.png)

**Comment by juristr** [comment]
@dean-g Here there is more info: https://github.com/storybookjs/storybook/issues/14909#issuecomment-842934111.

Basically you need to change the `optimization: true` property in the Angular app's configuration in `angular.json`. The `optimization` property can either be a boolean or an expanded object where you can enable/disable the settings for each individual optimization target. The link above shows how to fix it

**Comment by dean-g** [comment]
It's working now!. thank you very much @juristr

**Comment by literalpie** [comment]
Anyone who's having issues: Try with the latest storybook alpha (6.3.0-alpha.42) and `npm install @storybook/manager-webpack5`. This fixed all of the issues I was having

**Comment by dean-g** [comment]
I did the fresh run today. got into the following error: 

Cannot find module '@storybook/manager-webpack5'
Require stack:

installed: yarn add @storybook/manager-webpack5 

after this getting new error:

Cannot find module '../presets/manager-preset'
Require stack:

- node_modules/@storybook/manager-webpack5/dist/cjs/index.js
- node_modules/@storybook/core-server/dist/cjs/utils/get-manager-builder.js
- node_modules/@storybook/core-server/dist/cjs/build-static.js
- node_modules/@storybook/core-server/dist/cjs/index.js
- node_modules/@storybook/core/dist/cjs/server.js
- node_modules/@storybook/core/server.js
- node_modules/@nrwl/storybook/src/executors/storybook/storybook.impl.js
- node_modules/@nrwl/tao/src/shared/workspace.js
- node_modules/@nrwl/tao/src/commands/run.js
- node_modules/@nrwl/tao/index.js
- node_modules/@nrwl/cli/lib/run-cli.js

@juristr Can you please check this and advise?

**Comment by juristr** [comment]
@dean-g I was just about to test the release as well as this was the proposed solution from the meeting with the Storybook team last week.

Upgrading to the latest alpha release fixed the build for me as well. Here's my commit to the example repo: https://github.com/juristr/nx-angular12-storybook-webpack5/commit/dadfe1386da62b2a24b7e3af0e4dd10e620ce652

**Comment by dean-g** [comment]
Thank you @juristr 

My build got to succeed now but getting a few Conflicting errors and the storybook is not starting up. 

![image](https://user-images.githubusercontent.com/12157452/119574751-8f6f5f80-bd7b-11eb-9a05-cc6185387399.png)

Please check this and advise.

**Comment by dean-g** [comment]
It's working now after a restart. but still wondering the conflicting error

**Comment by dean-g** [comment]
@juristr Can you please help me with the following issue: [ng build failed](https://github.com/nrwl/nx/issues/5787)

**Comment by rp3e11** [comment]
I am facing the same DefinePlugin Conflicting values for process.env issue @dean-g had. A restart didn't help in my case. Did you change something else?

**Comment by rp3e11** [comment]
I created a reproduction [repo](https://github.com/rp3e11/nx-angular-tutorial) and to get the error exeucte the following
```
git checkout add-storybook
npm install
npx nx run ui:storybook
``
to reproduce the issue mentioned above. I used node 16.2 with npm 7.13 as well as node 14.17.0 with npm 6.14.13 and got the same result.

**Comment by juristr** [comment]
> I am facing the same DefinePlugin Conflicting values for process.env issue @dean-g had. A restart didn't help in my case. Did you change something else?

I saw that as well but need to investigate where that is coming from. Is it blocking for you? like does the serve or build not work?

**Comment by rp3e11** [comment]
Building and serving angular works, but storybook fails for angular v12 and storybook. Angular v11 works so I am not blocked in that sense, but I have to wait with upgrade until it is fixed or I can fix it.

**Comment by kingjordan** [comment]
Hello @juristr 

I can confirm that everything works for me (with the changes listed here: https://github.com/juristr/nx-angular12-storybook-webpack5/commit/dadfe1386da62b2a24b7e3af0e4dd10e620ce652) and nx 12.3.4 and angular 12.0.0 

On a side note, I have @ngneat/tailwind installed and it also works with that, the tailwind classes are getting picked up in storybook

I did have to remove the knobs registration with storybook because it is deprecated in favor of controls (storybook still runs but there is an error thrown for knobs if being used).  

Thanks so much

**Comment by jogelin** [comment]
Tried with `@storybook 6.3.0-beta.7` but it is failing because of looking for an `angular.json` file.

I am using a `workspace.json` file.

I thought it was fixed some time ago by https://github.com/storybookjs/storybook/pull/10881

**Comment by nzacca** [comment]
Hi @jogelin 

The issue was fixed by https://github.com/storybookjs/storybook/pull/10881 but it looks like that code has since been refactored into this: https://github.com/storybookjs/storybook/commit/4f004183eafc9be7cf404b6dc2b1a22801e86201#diff-563ad5c21cbbf4fe47dc95368b910f9d1eab0f507d7399e02da2bc957d570a38R10

The issue appears to be that this new method is using the angular-devkit's `readWorkspace` by default which now first [checks for the existence](https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/core/src/workspace/core.ts#L76) of `angular.json` and throws if it does not exist. Because of this, it never gets to the code [here](https://github.com/storybookjs/storybook/commit/4f004183eafc9be7cf404b6dc2b1a22801e86201#diff-563ad5c21cbbf4fe47dc95368b910f9d1eab0f507d7399e02da2bc957d570a38R31) that handles reading the `workspace.json` file.

**Comment by HarlesPilter** [comment]
If anyone is still having issues with this using the latest packages from storybook, angular, nx (In my case it was storybook - 6.3.0-beta.4, angular - 12.0.0, nx - 12.3.5). Then one thing you can try is to install @storybook/cli and add a script in your package.json to run storybook directly.
`"start:storybook": "start-storybook --config-dir libs/angular-components/.storybook"`
This way I got actual error messages in the console why the startup was failing (none of those errors were shown using the `nx run angular-components:storybook` command) In my case it was some paths in the config files were wrong.
@juristr Is it possible to get actual errors from storybook start to pass through the nx command?

**Comment by juristr** [comment]
🤔 hmmm that's interesting. I'll look into it. 
Also, if you folks happen to have some time, make sure to test the latest 6.3 betas
https://twitter.com/juristr/status/1400771614728466433?s=19

**Comment by harry-kalligeros** [comment]
I was able to make it work with the following configuration:

`
   "@nrwl/storybook": "12.3.6",
    "@nrwl/tao": "12.3.6",
    "@nrwl/web": "12.3.6",
    "@nrwl/workspace": "12.3.6",
    "@storybook/addon-actions": "6.3.0-beta.12",
    "@storybook/addon-console": "1.2.3",
    "@storybook/addon-controls": "6.3.0-beta.12",
    "@storybook/addon-knobs": "6.3.0-next.2",
    "@storybook/angular": "6.3.0-beta.12",
    "@storybook/builder-webpack5": "6.3.0-beta.12",
    "@storybook/manager-webpack5": "~6.3.0-beta.12",
`
For some reason, npm installed a react/react-dom version 15.x.x and I had an error in the console: `react.createContext is not a function`. When I realized that, I installed manually the versions 16.14.0 of these packages, and now I have storybook working.

**Comment by spierala** [comment]
Here I also have issues when running storybook with nrwl@12.3.6 and ng@12.
I created a new nx workspace from scratch, created an application and added storybook to that application as described [here](https://nx.dev/latest/angular/storybook/overview).

When running `nx run my-app:storybook` there is this error:

```
node_modules\webpack\lib\util\MapHelpers.js:17
        const value = map.get(key);
                          ^

TypeError: Cannot read property 'get' of undefined
```

**Comment by JSMike** [comment]
FYI, I'm also having success with Storybook 6.3.0-rc.2. Just make sure to update your builder, and convert all of your knobs to controls, because knobs are removed in 6.3.0. Heaviest lift is the knobs => controls but that was already deprecated, so we knew that was coming eventually. https://gist.github.com/shilman/8856ea1786dcd247139b47b270912324

**Comment by juristr** [comment]
> Heaviest lift is the knobs => controls but that was already deprecated, so we knew that was coming eventually.

@JSMike yeah we're already working to make sure the new setups get generated with controls rather than knobs and maybe we're able to provide a migration for that as well.

@spierala can you have a look at my setup: https://github.com/juristr/nx-angular12-storybook-webpack5. That's a working one, have a look especially at the last 2 commits which show the packages I've added to get it working.

I'm currently waiting for Storybook to cut the 6.3 and then we'll release an Nx migration to upgrade folks to a running version.

**Comment by stefan-schweiger** [comment]
@juristr when using your setup I'm still getting the `Conflicting values for 'process.env'`. I've even tried to update everything to the currently newest packages and setting `builder: 'webpack5'` explicitly but that also didn't fix it.

https://github.com/stefan-schweiger/nx-angular12-storybook-webpack5/commit/1684be031feb734ca6c8e8f6ef01b85a0d22bf17

**Comment by juristr** [comment]
@stefan-schweiger yep, that's a new thing that came up recently. It's being discussed here on the Storybook repo and a blocker for their 6.3 release atm
https://github.com/storybookjs/storybook/issues/14257

**Comment by ahmedbillalimtiazghumman** [comment]
<img width="1164" alt="Screenshot 2021-06-16 at 7 21 17 PM" src="https://user-images.githubusercontent.com/55593620/122236828-48870c80-ced8-11eb-95c2-3e19cf2ce7e8.png">

Is there anyone who can help me resolve this error related to build angular-storybook.

**Comment by ahmedbilal1994** [comment]
> <img alt="Screenshot 2021-06-16 at 7 21 17 PM" width="1164" src="https://user-images.githubusercontent.com/55593620/122236828-48870c80-ced8-11eb-95c2-3e19cf2ce7e8.png">
> 
> Is there anyone who can help me resolve this error related to build angular-storybook.

@juristr

**Comment by juristr** [comment]
@ahmedbillalimtiazghumman hard to tell without more info. Looks to me like the `webpack.config.js` in your `.storybook` folder might have something in there which is no more valid for Webpack 5. The error in the screenshot you provided shows some hints about the error

**Comment by ahmedbillalimtiazghumman** [comment]
@juristr i have not changed anything, it is working with angular v9, i am getting this error after updating my angular v12.

**Comment by thutinavaneethreddy** [comment]
If someone is still looking around for a solution, here is what worked for me with storybook 6.2.7 version :

npm install webpack@5.36.2 --save-dev
npm install dotenv-webpack@6.0.4 --save-dev (This fixed "cannot read property 'get' of undefined" but gave a new issue about snapshots)
npm install html-webpack-plugin@5.3.1 --save-dev (This fixed snapshot issue)
In the build config in angular.json, change "optimization" : "false" to
"optimization": {
"styles": false,
"scripts": false,
"fonts": false
}

This fixed the warning - "Cannot read property 'minify' of undefined"

After these, I was able to build and run storybook without any errors

**Comment by markhus-aurelius** [comment]
> @juristr i have not changed anything, it is working with angular v9, i am getting this error after updating to angular v12.

That's exactly his point. You HAVE to change something to make your config compatible with Webpack 5

**Comment by shagene** [comment]
Would like to know myself what needs to be changed to work with Angular 12 / Webpack 5

**Comment by stefan-schweiger** [comment]
@shagene my current workaround is to add the following:

`/.storybook/main.js`:
```
builder: 'webpack5',
// I don't know if the config below here is needed, but I don't see why not to add it
angularOptions: {
  enableIvy: true,
},
addons: [
  'storybook-addon-angular-ivy',
  // ...
]
```

`/.storybook/webpack.config.js`:
```
module.exports = async ({ config, mode }) => {
  // Make whatever fine-grained changes you need
  config.plugins = [
    ...config.plugins.filter((plugin) => !plugin.definitions),
    new webpack.DefinePlugin({
      'process.env.NODE_ENV': JSON.stringify(
        mode === 'DEVELOPMENT' ? 'development' : 'production'
      ),
      'process.env.NODE_PATH': JSON.stringify([]),
      'process.env.STORYBOOK': JSON.stringify('true'),
      'process.env.PUBLIC_PATH': JSON.stringify('.'),
    }),
  ];

  // Return the altered config
  return config;
};
```

I still get the warning in the console, but at least storybook starts up and I can use it.

**Comment by juristr** [comment]
> @juristr Is it possible to get actual errors from storybook-start to pass through the nx command?

@HarlesPilter This PR should fix that https://github.com/nrwl/nx/pull/6129

**Comment by jogelin** [comment]
@juristr do you know if we Nx will integrate the new storybook builders? https://github.com/storybookjs/storybook/pull/15061

**Comment by juristr** [comment]
@jogelin hey, it's on my list to give them a look and verify whether it might be useful integrating them into Nx (once I'm back from vacation 😉)

**Comment by udos86** [comment]
Is this officially fixed? 

Sorry for asking, but unfortunately, I still can't get Angular libraries to work with Storybook out of the box:

### Steps to Reproduce

1. `npx create-nx-workspace --cli=angular` (create empty workspace)
2. `npm i @nrwl/angular @nrwl/storybook -D` (install Nrwl packages)
3. `ng g @nrwl/angular:lib demo-lib --publishable --strict --importPath @udos86/demo-lib` (create sample lib)
4. `ng g c button --project demo-lib` (create sample component)
5. `ng g @nrwl/angular:storybook-configuration demo-lib` (create storybook configuration for sample lib)
6. `ng run demo-lib:storybook --verbose` (run storybook)


### Failure Logs
```cmd
> nx run demo-lib:storybook
info => Loading presets
info => Loading 1 config file in "D:\dev\nx-sb-cli-ng-empty\libs\demo-lib\.storybook"
info => Loading 8 other files in "D:\dev\nx-sb-cli-ng-empty\libs\demo-lib\.storybook"
info => Adding stories defined in "D:\dev\nx-sb-cli-ng-empty\libs\demo-lib\.storybook\main.js"
info => Found custom tsconfig.json
info => Using implicit CSS loaders
info => Loading angular-cli config
info => Using angular project "demo-lib:build" for configuring Storybook
ERR! => Could not get angular cli webpack config
TypeError: s.startsWith is not a function
    at Object.error (D:\dev\nx-sb-cli-ng-empty\node_modules\@nrwl\tao\src\shared\logger.js:10:15)
    at D:\dev\nx-sb-cli-ng-empty\node_modules\@nrwl\storybook\src\executors\storybook\storybook.impl.js:43:33
```

### Environment
```cmd
  Node : 14.16.1
  OS   : win32 x64
  npm  : 6.14.12

  nx : Not Found
  @nrwl/angular : 12.5.6
  @nrwl/cli : 12.5.6
  @nrwl/cypress : 12.5.6
  @nrwl/devkit : 12.5.6
  @nrwl/eslint-plugin-nx : 12.5.6
  @nrwl/express : Not Found
  @nrwl/jest : 12.5.6
  @nrwl/linter : 12.5.6
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 12.5.6
  @nrwl/web : Not Found
  @nrwl/workspace : 12.5.6
  @nrwl/storybook : 12.5.6
  @nrwl/gatsby : Not Found
  typescript : 4.2.4
```
